File: Appendix Linear Algebra/Dirichlet Energy/Dirichlet Energy.py
#### Dirichlet Energy

import numpy as np

A = np.array([[-1,-1, 0, 1, 0],
              [ 1, 0,-1, 0, 0],
              [ 0, 0, 1,-1,-1],
              [ 0, 1, 0, 0, 1]])

VSmooth = np.array([1, 2, 2, 1])
VRough  = np.array([1,-1, 2,-1])

nSmooth = np.linalg.norm(np.dot(A.T,VSmooth))**2
nRough  = np.linalg.norm(np.dot(A.T,VRough))**2

print(nSmooth,"\n",nRough)
    # 2.9999999999999996 
    # 27.0

File: Appendix Linear Algebra/Dirichlet Energy/Dirichlet Energy.jl
#### Dirichlet Energy

using LinearAlgebra

A = [-1 -1  0  1  0;
      1  0 -1  0  0;
      0  0  1 -1 -1;
      0  1  0  0  1]

VSmooth = [1, 2, 2, 1]
VRough  = [1,-1, 2,-1]

nSmooth = norm(A'*VSmooth)^2
nRough  = norm(A'*VRough)^2

println(nSmooth,"\n",nRough)
      # 2.9999999999999996
      # 27.0